x/leverage/keeper/keeper.go

```diff
@@ -346,226 +346,76 @@ func (k Keeper) Decollateralize(ctx sdk.Context, borrowerAddr sdk.AccAddress, co
-			"borrower doesn't have %s as a collateral", desiredReward.Denom)
-	}
-
-	// reward amount cannot exceed available collateral
-	if reward.Amount.GT(maxReward) {
-		// reduce repayment.Amount to the maximum value permitted by the available collateral reward
-		repayment.Amount = repayment.Amount.Mul(maxReward).Quo(reward.Amount)
-		reward.Amount = maxReward
+		return sdk.Coin{}, sdk.Coin{}, sdk.Coin{}, err
 	}
 
-	// final check for invalid liquidation (negative/zero value after reductions above)
-	if !repayment.Amount.IsPositive() {
-		return sdk.ZeroInt(), sdk.ZeroInt(), types.ErrInvalidAsset.Wrap(repayment.String())
-	}
-
-	if desiredReward.Amount.IsPositive() {
-		// user-controlled minimum ratio of reward to repayment, expressed in collateral base assets (not uTokens)
-		rewardTokenEquivalent, err := k.ExchangeUToken(ctx, reward)
-		if err != nil {
-			return sdk.ZeroInt(), sdk.ZeroInt(), err
-		}
-
-		minimumRewardRatio := sdk.NewDecFromInt(desiredReward.Amount).QuoInt(desiredRepayment.Amount)
-		actualRewardRatio := sdk.NewDecFromInt(rewardTokenEquivalent.Amount).QuoInt(repayment.Amount)
-		if actualRewardRatio.LT(minimumRewardRatio) {
-			return sdk.ZeroInt(), sdk.ZeroInt(), types.ErrLiquidationRewardRatio
-		}
-	}
-
-	// send repayment to leverage module account
-	if err = k.bankKeeper.SendCoinsFromAccountToModule(
-		ctx, liquidatorAddr,
-		types.ModuleName,
-		sdk.NewCoins(repayment),
-	); err != nil {
-		return sdk.ZeroInt(), sdk.ZeroInt(), err
-	}
-
-	// update the remaining borrowed amount
-	owed := borrowed.AmountOf(repayment.Denom).Sub(repayment.Amount)
-	if err = k.setBorrow(ctx, borrowerAddr, sdk.NewCoin(repayment.Denom, owed)); err != nil {
-		return sdk.ZeroInt(), sdk.ZeroInt(), err
-	}
-
-	// Reduce borrower collateral by reward amount
-	newBorrowerCollateral := sdk.NewCoin(reward.Denom, maxReward.Sub(reward.Amount))
-	if err = k.setCollateralAmount(ctx, borrowerAddr, newBorrowerCollateral); err != nil {
-		return sdk.ZeroInt(), sdk.ZeroInt(), err
-	}
-
-	// Send rewards to liquidator's account.
-	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, liquidatorAddr, sdk.NewCoins(reward))
-	if err != nil {
-		return sdk.ZeroInt(), sdk.ZeroInt(), err
-	}
-
-	// Detect bad debt (collateral == 0 after reward) for repayment by protocol reserves (see ADR-004)
-	if collateral.Sub(sdk.NewCoins(reward)).IsZero() {
-		for _, coin := range borrowed {
-			// Mark repayment denom as bad debt only if some debt remains after
-			// this liquidation. All other borrowed denoms were definitely not
-			// repaid in this liquidation so they are always marked as bad debt.
-			if coin.Denom != repayment.Denom || owed.IsPositive() {
-				if err := k.setBadDebtAddress(ctx, borrowerAddr, coin.Denom, true); err != nil {
-					return sdk.ZeroInt(), sdk.ZeroInt(), err
-				}
+	// detect bad debt if collateral is completely exhausted
+	if k.GetBorrowerCollateral(ctx, borrowerAddr).IsZero() {
+		// TODO: exclude blacklisted collateral
```

Comment:
Is there a task for that? if not then let's add it and include in the comment

Reply:
Added this in a commit this morning